business/openid_auth.go

```diff
@@ -28,8 +38,141 @@ type OpenIdMetadata struct {
 	ResponseTypesSupported []string `json:"response_types_supported"`
 }
 
+type OpenIdCallbackParams struct {
+	Code          string
+	ExpiresOn     time.Time
+	IdToken       string
+	Nonce         string
+	NonceHash     []byte
+	ParsedIdToken *jwt.Token
+	State         string
+	Subject       string
+}
+
 var cachedOpenIdMetadata *OpenIdMetadata
 
+func BuildOpenIdJwtClaims(openIdParams *OpenIdCallbackParams) *config.IanaClaims {
+	return &config.IanaClaims{
+		SessionId: openIdParams.IdToken,
+		StandardClaims: jwt.StandardClaims{
+			Subject:   openIdParams.Subject,
+			ExpiresAt: openIdParams.ExpiresOn.Unix(),
+			Issuer:    config.AuthStrategyOpenIdIssuer,
+		},
+	}
+}
+
+func ExtractOpenIdCallbackParams(w http.ResponseWriter, r *http.Request) (params *OpenIdCallbackParams, err error) {
+	params = &OpenIdCallbackParams{}
+
+	// Get the nonce code hash
+	var nonceCookie *http.Cookie
+	if nonceCookie, err = r.Cookie(OpenIdNonceCookieName); err == nil {
+		params.Nonce = nonceCookie.Value
+
+		hash := sha256.Sum224([]byte(nonceCookie.Value))
+		params.NonceHash = make([]byte, sha256.Size224)
+		copy(params.NonceHash, hash[:])
+	}
+
+	// Parse/fetch received form data
+	err = r.ParseForm()
+	if err != nil {
+		err = fmt.Errorf("error parsing form info: %w", err)
+	} else {
+		// Read relevant form data parameters
+		params.Code = r.Form.Get("code")
+		params.IdToken = r.Form.Get("id_token")
+		params.State = r.Form.Get("state")
+	}
+
+	// Cleanup: Delete the nonce cookie since we no longer need it.
+	deleteNonceCookie := http.Cookie{
+		Name:     OpenIdNonceCookieName,
+		Expires:  time.Unix(0, 0),
+		HttpOnly: true,
+		Path:     config.Get().Server.WebRoot,
+		SameSite: http.SameSiteStrictMode,
+		Value:    "",
+	}
+	http.SetCookie(w, &deleteNonceCookie)
+
+	return
+}
+
+//func CheckOpenIdImplicitFlowParams(params *OpenIdCallbackParams) string {
+//	if params.NonceHash == nil {
+//		return "No nonce code present. Login window timed out."
+//	}
+//	if params.State == "" {
+//		return "State parameter is empty or invalid."
+//	}
+//	if params.IdToken == "" {
+//		return "Token is empty or invalid."
+//	}
+//
+//	return ""
+//}
```

Comment:
We should remove this commented code if it will never be used.

Reply:
Unused right now, but it's here so that I don't forget it.
All these new functions are duplicating code inside the `performOpenIdAuthentication`. The code flow is more or less the same in both "openid flows", but order of execution changes. So, in a follow up PR, my idea is to change `performOpenIdAuthentication` to unify code and de-duplicate.  So, this commented code will be un-commented :)